ambry-quota/src/main/java/com/github/ambry/quota/QuotaResourceSynchronizer.java

```diff
@@ -0,0 +1,133 @@
+    this.quotaMetrics = quotaMetrics;
+  }
+
+  /**
+   * Acquires the lock for the quota resource represented by the specified {@link QuotaResource}.
+   * @param quotaResource {@link QuotaResource} object representing the quota resource for which lock needs to be
+   *                                           acquired.
+   */
+  public void lock(QuotaResource quotaResource) {
+    quotaResourceMutexCache.get(quotaResource).lock();
+  }
+
+  /**
+   * Releases the lock for the quota resource represented by the specified {@link QuotaResource}.
+   * @param quotaResource {@link QuotaResource} object representing the quota resource for which lock needs to be
+   *                      released.
+   */
+  public void unlock(QuotaResource quotaResource) {
+    ReentrantLock lock = quotaResourceMutexCache.mutexCache.getOrDefault(quotaResource, null);
+    if (lock != null) {
+      lock.unlock();
+    } else {
+      LOGGER.warn("Attempted unlock for {} when lock wasn't held.", quotaResource);
+      quotaMetrics.quotaResourceInvalidUnlockAttemptCount.inc();
+    }
+  }
+
+  /**
+   * Tear down and cleanup the mutex provider.
+   */
+  public void tearDown() {
+    quotaResourceMutexCache.cleanupService.shutdown();
+  }
+
+  /**
+   * A cache of unique mutexes for all {@link QuotaResource} objects that represent the same quota resource.
+   */
+  static class QuotaResourceMutexCache {
+    private static final String CLEANUP_THREAD_NAME_PREFIX = "quota-resource-mutex-cache-cleaner";
+    private static final int CLEANUP_PERIOD_IN_MINS = 10;
+    protected final ConcurrentHashMap<QuotaResource, ReentrantLock> mutexCache = new ConcurrentHashMap<>();
+    private final ReentrantReadWriteLock cleanupLock = new ReentrantReadWriteLock();
+    private final ScheduledExecutorService cleanupService;
+
+    /**
+     * Constructor for {@link QuotaResourceSynchronizer}.
+     */
+    protected QuotaResourceMutexCache() {
+      cleanupService = Utils.newScheduler(1, CLEANUP_THREAD_NAME_PREFIX, true);
+      cleanupService.scheduleAtFixedRate(this::cleanupCache, CLEANUP_PERIOD_IN_MINS, CLEANUP_PERIOD_IN_MINS,
+          TimeUnit.MINUTES);
+    }
+
+    /**
+     * Returns the mutex in the form of {@link ReentrantLock} object for the specified {@link QuotaResource}.
+     * If the mutex is not already present in the cache, it creates a new mapping.
+     * @param quotaResource {@link QuotaResource} object.
+     * @return ReentrantLock object as mutex for the specified quotaResource.
+     */
+    ReentrantLock get(QuotaResource quotaResource) {
+      try {
+        cleanupLock.readLock().lock();
+        mutexCache.putIfAbsent(quotaResource, new ReentrantLock());
+        return mutexCache.get(quotaResource);
+      } finally {
+        cleanupLock.readLock().unlock();
+      }
+    }
+
+    /**
+     * Method to cleanup the mutexes that are no longer in use.
+     */
+    protected void cleanupCache() {
+      try {
+        cleanupLock.writeLock().lock();
```

Comment:
Concurrent iteration is supported in ConcurrentHashMap. My understanding of the RW lock here is to account for the case where someone is getting a lock but has not locked it yet and this bg thread sees that no one is using the lock and removes it. Then, another thread tries to get a lock for that resource, but it actually gets a new lock. Is this correct?

Here i'm not even sure if that will fully protect it since the ReentrantLock::lock is called outside of the cleanupLock read scope.